File: algorithm/leetcodeObjectiveC/balanced-binary-tree.m
// 110. Balanced Binary Tree
// Difficulty: Easy

// Given a binary tree, determine if it is height-balanced.

// For this problem, a height-balanced binary tree is defined as a binary tree
// in which the depth of the two subtrees of every node never differ by more than 1.

// Time:  O(n)
// Space: O(h)

#import <Foundation/Foundation.h>

@interface TreeNode : NSObject

@property (atomic, assign) NSInteger value;
@property (atomic, strong) TreeNode* left;
@property (atomic, strong) TreeNode* right;

-(id)initWithValue:(NSInteger)value leftNode:(TreeNode*)left rightNode:(TreeNode*)right;
-(id)initWithValue:(NSInteger)value leftNode:(TreeNode *)left;
-(id)initWithValue:(NSInteger)value rightNode:(TreeNode *)right;
-(id)initWithValue:(NSInteger)value;
-(id)init;

@end

@implementation TreeNode

-(id)initWithValue:(NSInteger)value leftNode:(TreeNode*)left rightNode:(TreeNode*)right {
  self = [super init];
  if (self) {
    self.value = value;
    self.left = left;
    self.right = right;
  }
  return self;
}

-(id)initWithValue:(NSInteger)value leftNode:(TreeNode *)left {
  return [self initWithValue:value leftNode:left rightNode:nil];
}

-(id)initWithValue:(NSInteger)value rightNode:(TreeNode *)right {
  return [self initWithValue:value leftNode:nil rightNode:right];
}

-(id)initWithValue:(NSInteger)value {
  return [self initWithValue:value leftNode:nil rightNode:nil];
}

-(id)init {
  return [self initWithValue:0 leftNode:nil rightNode:nil];
}

@end

// return the height of `root` if `root` is a balanced tree,
// else, return -1.

int balancedHeight (TreeNode* root) {
  if (!root) {
    return 0;
  }
  int left = balancedHeight(root.left);
  int right = balancedHeight(root.right);
  if (left < 0 || right < 0 || abs(left - right) > 1) {
    return -1;
  }
  return MAX(left, right) + 1;
}

BOOL isBalanced (TreeNode* root) {
  return balancedHeight(root) >= 0;
}

File: data-structure-and-algorithm/typescript/leetcode/src/balanced-binary-tree.ts
/**
Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined as:

a binary tree in which the left and right subtrees of every node differ in height by no more than 1.



Example 1:


Input: root = [3,9,20,null,null,15,7]
Output: true
Example 2:


Input: root = [1,2,2,3,3,null,null,4,4]
Output: false
Example 3:

Input: root = []
Output: true


Constraints:

The number of nodes in the tree is in the range [0, 5000].
-104 <= Node.val <= 104
*/

function isBalanced(root) {
  return isBalancedDFS(root) >= 0;
}

function isBalancedDFS(root) {
  if (root === null) {
    return 0;
  }

  const left = isBalancedDFS(root.left);
  const right = isBalancedDFS(root.right);
  if (left < 0 || right < 0 || Math.abs(right - left) > 1) {
    return -1;
  }

  return Math.max(left, right) + 1;
}
